zerver/models.py

```diff
@@ -1205,6 +1205,22 @@ def get_url(self):
         # type: () -> text_type
         return u"/user_uploads/%s" % (self.path_id)
 
+def validate_attachment_request(user_profile, path_id):
+    try:
+        attachment = Attachment.objects.get(path_id=path_id)
+        messages = attachment.messages.all()
+
+        if user_profile == attachment.owner:
+            return True
+        elif attachment.is_realm_public and attachment.realm == user_profile.realm:
+            return True
+        elif UserMessage.objects.filter(user_profile=user_profile, message__in=messages).exists():
+            return True
```

Comment:
I don't see a test that covers this case; to add one, you'll want to send an attachment in a cross-realm private message, and confirm the recipient can view it.  The `test_to_zulip_realm` test in test_messages.py has example code for how to construct a cross-realm private message.